File: leetcode-js/117-populating-next-right-pointers-in-each-node-ii.js
/**
 * Definition for binary tree with next pointer.
 * function TreeLinkNode(val) {
 *     this.val = val;
 *     this.left = this.right = this.next = null;
 * }
 */

/**
 * @param {TreeLinkNode} root
 * @return {void} Do not return anything, modify tree in-place instead.
 */
var connect = function(root) {
    if (!root) {
        return;
    }

    let tl = root;

    while (tl) {
        let tc = tl;
        let l = null;
        let c = null;
        while (tc) {
            if (tc.left) {
                if (!l) {
                    l = tc.left;
                    c = l;
                } else {
                    c.next = tc.left;
                    c = c.next;
                }
            }
            if (tc.right) {
                if (!l) {
                    l = tc.right;
                    c = l;
                } else {
                    c.next = tc.right;
                    c = c.next;
                }
            }
            tc = tc.next;
        }
        tl = l;
    }
};

File: leetcode-java/117-populating-next-right-pointers-in-each-node-ii.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void connect(TreeLinkNode root) {
        TreeLinkNode topMostLeft = root;

        while (topMostLeft != null) {
            TreeLinkNode topCurrent = topMostLeft;
            TreeLinkNode mostLeft = null;
            TreeLinkNode current = null;

            while (topCurrent != null) {
                if (topCurrent.left != null) {
                    if (mostLeft == null) {
                        mostLeft = topCurrent.left;
                        current = mostLeft;
                    } else {
                        current.next = topCurrent.left;
                        current = current.next;
                    }
                }

                if (topCurrent.right != null) {
                    if (mostLeft == null) {
                        mostLeft = topCurrent.right;
                        current = mostLeft;
                    } else {
                        current.next = topCurrent.right;
                        current = current.next;
                    }
                }

                topCurrent = topCurrent.next;
            }

            topMostLeft = mostLeft;
        }
    }
}
